src/matrix.rs

```diff
@@ -0,0 +1,152 @@
+use curl::easy::Easy;
+use regex::Regex;
+use serde::*;
+use serde::{
+	Deserialize,
+	Serialize,
+};
+use snafu::ResultExt;
+use std::io::{
+	stdout,
+	Write,
+};
+
+use crate::{
+	error,
+	Result,
+};
+
+#[derive(Deserialize, Debug)]
+pub struct LoginResponse {
+	pub access_token: String,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct CreateRoomResponse {
+	pub room_id: String,
+}
+
+pub fn login(homeserver: &str, username: &str, password: &str) -> Result<LoginResponse> {
+	let mut dst = Vec::new();
+	let mut handle = Easy::new();
+	handle
+		.url(format!("{}/_matrix/client/r0/login", homeserver).as_ref())
+		.or_else(error::map_curl_error)?;
+	handle
+                .post_fields_copy(
+                        format!(
+                                "{{\"type\":\"m.login.password\", \"identifier\": {{ \"type\": \"m.id.thirdparty\", \"medium\": \"email\", \"address\": \"{}\" }}, \"password\":\"{}\"}}",
```

Comment:
Could this use the `serde_json::json!` macro, and converted into a string rather than using a string directly?